SilKit/source/core/vasio/RingBuffer.hpp

```diff
@@ -0,0 +1,352 @@
+    _size -= numBytes;
+
+    // size checks
+    SizeCheck();
+}
+
+template <class T>
+bool RingBuffer<T>::Write(const T& elem)
+{
+    if (Full()) // we don't allow overwriting
+    {
+        return false;
+    }
+
+    _buffer.at(_wPos) = elem;
+    AdvanceWPos(1);
+
+    return true;
+}
+
+template <class T>
+bool RingBuffer<T>::Read(T& elem)
+{
+    if (Empty()) // no data to read
+    {
+        return false;
+    }
+
+    elem = _buffer.at(_rPos);
+    AdvanceRPos(1);
+
+    return true;
+}
+
+template <class T>
+bool RingBuffer<T>::Write(const std::vector<T>& elem)
+{
+    // we don't allow overwriting
+    if (elem.size() > (_capacity - _size))
+    {
+        return false;
+    }
+
+    // copy data to first contiguous array
+    size_t numBytesArrayOne = std::min((_capacity - _wPos), elem.size());
+    std::memcpy(_buffer.data() + _wPos, elem.data(), numBytesArrayOne);
+
+    // copy data to second contiguous array (if necessary)
+    if (numBytesArrayOne < elem.size())
+    {
+        std::memcpy(_buffer.data(), elem.data() + numBytesArrayOne, elem.size() - numBytesArrayOne);
+    }
+
+    AdvanceWPos(elem.size());
+
+    return true;
+}
+
+template <class T>
+bool RingBuffer<T>::Read(std::vector<T>& elem, bool advanceRPos)
+{
+    // make sure, we only copy as many bytes as are contained in the buffer
+    if (elem.size() > _size)
+    {
+        return false;
+    }
+
+    // copy data from first contiguous array
+    size_t numBytesArrayOne = std::min((_capacity - _rPos), elem.size());
+    std::memcpy(elem.data(), _buffer.data() + _rPos, numBytesArrayOne);
+
+    // copy data from second contiguous array (if necessary)
+    if (numBytesArrayOne < elem.size())
+    {
+        std::memcpy(elem.data() + numBytesArrayOne, _buffer.data(), elem.size() - numBytesArrayOne);
+    }
+
+    if (advanceRPos)
+    {
+        AdvanceRPos(elem.size());
+    }
+
+    return true;
+}
+
+template <class T>
+size_t RingBuffer<T>::GetSizeArrayOne()
+{
+    size_t arrayOneSize = std::min(_capacity - _wPos, _capacity - _size);
+
+    // handle edge cases
+    if (Empty())
+    {
+        arrayOneSize = _capacity - _wPos;
+    }
+
+    return arrayOneSize;
+}
+
+template <class T>
+size_t RingBuffer<T>::GetSizeArrayTwo()
+{
+    size_t arrayTwoSize = (_capacity - _size) - GetSizeArrayOne();
+
+    // handle edge cases
+    if (Empty())
+    {
+        arrayTwoSize = _wPos;
+    }
+
+    return arrayTwoSize;
+}
+
+// get first contiguous array contained in _buffer (free slots)
+template <class T>
+auto RingBuffer<T>::GetArrayOne() -> RingBuffer<T>::BufArray
+{
+    return BufArray{_buffer.data() + _wPos, GetSizeArrayOne()};
+}
+
+// get second contiguous array contained in _buffer (free slots)
+template <class T>
+auto RingBuffer<T>::GetArrayTwo() -> RingBuffer<T>::BufArray
+{
+    return BufArray{_buffer.data(), GetSizeArrayTwo()};
+}
+
+template <class T>
+size_t RingBuffer<T>::Capacity()
+{
+    return _capacity;
```

Comment:
If we try to mimic the implementation of std::containers it should actually return 
`return _buffer.capacity()` since capacity is usually about the allocated size, not the numbers of elements the buffer currently holds.

Reply:
Yes and no, the capacity of the underlying buffer should always be it's size, since this class uses `resize` only to make the buffer bigger. Also the 'capacity' of the ring-buffer must be allocated already.

Reply:
Ah yes, got it. Then we need to use `size()` here, agreed 👍  Thanks for the explanation.

Reply:
Thanks for the hint! I just realized that _capacity is actually redundant.